### ISSUE TITLE
Performance issues with linear interpolations

### ISSUE THREAD

**Issue by angel-devicente** [JuliaMath/Interpolations.jl#311]
Hi,

as suggested by Tim Holy in the Discourse page, I send here the details of the issue I'm having with Interpolations.jl.

I have four versions of a test function to do some interpolations:

```
function interp_test1(size)
    xs = range(1,size,length=size)
    ys = range(1,size,length=size)
    zs = range(1,size,length=size)

    xsi = xs .+ 0.5 ; ysi = ys .+ 0.5 ; zsi = zs .+ 0.5

    dat = randn(size,size,size)

    dat_out = zeros(size,size,size)
    dat_int = LinearInterpolation((xs,ys,zs),dat,extrapolation_bc = Line())  # ~23s 124M allocs 4.65 GiB                                                                                                             
    for C in CartesianIndices((length(xsi),length(ysi),length(zsi)))
        dat_out[C] = dat_int(xsi[C[1]],ysi[C[2]],zsi[C[3]])
    end

    return sum(dat_out)
end


function interp_test2(size)
    xs = range(1,size,length=size)
    ys = range(1,size,length=size)
    zs = range(1,size,length=size)

    xsi = xs .+ 0.5 ; ysi = ys .+ 0.5 ; zsi = zs .+ 0.5

    dat = randn(size,size,size)

    dat_out = zeros(size,size,size)
    dat_int = extrapolate(interpolate(dat, BSpline(Linear())), Line())        # ~16s 125M allocs 4.65 GiB                                                                                                            
    for C in CartesianIndices((length(xsi),length(ysi),length(zsi)))
        dat_out[C] = dat_int(xsi[C[1]],ysi[C[2]],zsi[C[3]])
    end

    return sum(dat_out)
end

function interp_test3(size)
    xs = range(1,size,length=size)
    ys = range(1,size,length=size)
    zs = range(1,size,length=size)

    xsi = xs .+ 0.5 ; ysi = ys .+ 0.5 ; zsi = zs .+ 0.5

    dat = randn(size,size,size)

    dat_out = zeros(size-1,size-1,size-1)
    dat_int = LinearInterpolation((xs,ys,zs),dat)     # ~20s 124M allocs 4.64 GiB                                                                                                                                    

    for C in CartesianIndices((length(xsi)-1,length(ysi)-1,length(zsi)-1))
        dat_out[C] = dat_int(xsi[C[1]],ysi[C[2]],zsi[C[3]])
    end

    return sum(dat_out)
end


function interp_test4(size)
    xs = range(1,size,length=size)
    ys = range(1,size,length=size)
    zs = range(1,size,length=size)

    xsi = xs .+ 0.5 ; ysi = ys .+ 0.5 ; zsi = zs .+ 0.5

    dat = randn(size,size,size)

    dat_out = zeros(size-1,size-1,size-1)
    dat_int = interpolate(dat,BSpline(Linear()))       # ~7s  24 allocs   2.79 GiB                                                                                                                                   

    for C in CartesianIndices((length(xsi)-1,length(ysi)-1,length(zsi)-1))
        dat_out[C] = dat_int(xsi[C[1]],ysi[C[2]],zsi[C[3]])
    end

    return sum(dat_out)
end
```

The difference with the four versions are:
- test1 : uses extrapolation and ranges to the interpolations routine
- test2 : extrapolation, but no ranges
- test3: no extrapolation, but uses ranges
- test4: no extrapolation and no ranges

From running these versions, you can see that all the versions have huge allocation numbers, except for test4, but actually I need test1 :-) 

```
julia> @time interp_test1(500)                                                                                                                                                                                       
 23.746982 seconds (125.00 M allocations: 4.657 GiB, 1.32% gc time)                                                                                                                                                  
7913.71601587713                                                                                                                                                                                                     
                                                                                                                                                                                                                     
julia> @time interp_test2(500)                                                                                                                                                                                       
 16.514156 seconds (125.00 M allocations: 4.657 GiB, 1.92% gc time)                                                                                                                                                  
-416.88550691519686                                                                                                                                                                                                  
                                                                                                                                                                                                                     
julia> @time interp_test3(500)                                                                                                                                                                                       
 20.049974 seconds (124.25 M allocations: 4.640 GiB, 1.55% gc time)                                                                                                                                                  
16929.240370912485                                                                                                                                                                                                   
                                                                                                                                                                                                                     
julia> @time interp_test4(500)                                                                                                                                                                                       
  7.976706 seconds (14 allocations: 2.788 GiB, 1.96% gc time)                                                                                                                                                        
17234.593944441876

```
In case it can be useful, I upload a zip file with the Profile and ProfileView generated data for each of the tests.


[profs.zip](https://github.com/JuliaMath/Interpolations.jl/files/3008492/profs.zip)

**Comment by tpapp** [comment]
I may have run into the same or a similar problem. Below is an MWE that demonstrates this, on 
```julia
julia> versioninfo()
Julia Version 1.2.0-pre.0
Commit 8a84ba5018 (2019-04-11 05:23 UTC)
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: Intel(R) Core(TM) i7-6560U CPU @ 2.20GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-6.0.1 (ORCJIT, skylake)
Environment:
  JULIA_CMDSTAN_HOME = /home/tamas/src/cmdstan-2.18.1/

(v1) pkg> st Interpolations
    Status `~/.julia/environments/v1/Project.toml`
  [a98d9a8b] Interpolations v0.11.2+ [`~/.julia/dev/Interpolations`]
  [6fe1bfb0] OffsetArrays v0.10.0+ #master (https://github.com/JuliaArrays/OffsetArrays.jl.git)
  [90137ffa] StaticArrays v0.10.3
  [37e2e46d] LinearAlgebra 
```

In the MWE, I use ranges for the last two non-interpolated indices. This was key for me to reproduce the problem, as it goes away without it.

```julia
using Interpolations, BenchmarkTools

#### test grid and arrays (we will hardcode dimensions)

a_grid = 0:0.5:5
A = [a1 + a2 + j + o for a1 in a_grid, a2 in a_grid, j in 1:2, o in 1:3]
a = (2.1, 3.8)

#### using Interpolations

I = interpolate((a_grid, a_grid, 1:2, 1:3),
                A, (Gridded(Linear()), Gridded(Linear()), NoInterp(), NoInterp()))

evalI(I, a) = I(a..., 1:2, 1:3)

@btime evalI($I, $a) # 3.022 μs (23 allocations: 17.89 KiB)

#### naive implementation

struct LinInt{T}
    ix::Int
    α::T
end

function lin_int(grid, x)       # linear interpolation
    i = searchsortedlast(grid, x)
    @assert i > 0
    l, r = grid[i], grid[i + 1]
    LinInt(i, (x - l) / (r - l))
end

lin_int(::Nothing, x::Integer) = x # no interpolation

function _itpG(A, before, li::LinInt, after...)
    @inbounds ((1 - li.α) * _itpG(A, (before..., li.ix), after...) +
               li.α * _itpG(A, (before..., li.ix + 1), after...))
end

_itpG(A, before, i::Integer, after...) = _itpG(A, (before..., i), after...)

_itpG(A, before) = A[before...]

itpG(A, ixs...) = _itpG(A, (), ixs...)

function evalG(a_grid, A, a1, a2)
    li1, li2 = lin_int(a_grid, a1), lin_int(a_grid, a2)
    [itpG(A, li1, li2, j, o) for j in 1:2, o in 1:3]
end

@btime evalG($a_grid, $A, $a...) # 168.572 ns (3 allocations: 224 bytes)
```

**Comment by timholy** [comment]
One small correction, any time you use `LinearInterpolation` you're using extrapolation. If you really want to avoid it you need to use the original interface.

Fix in #315.
